node/src/effect.rs

```diff
@@ -1060,10 +1061,19 @@ impl<REv> EffectBuilder<REv> {
     #[allow(dead_code)]
     pub(crate) async fn get_validators(
         self,
-        _root_hash: Digest,
-        _era: u64,
-    ) -> Result<Option<HashMap<PublicKey, u64>>, execution::Error> {
-        todo!("get_validators")
+        get_request: GetEraValidatorsRequest,
+    ) -> Result<Option<ValidatorWeights>, GetEraValidatorsError>
```

Comment:
I wonder... Why is the `Ok()` variant an `Option`? Isn't being unable to return a set of validators an error condition? I see that the stub before your changes was defined like that as well, so I'm probably just missing something.

Reply:
Yea, that's convention we've had on EE side for some time which I've forgotten in earlier commits. Lack of value is not the same as error

Reply:
Hmm, I understand that sometimes a lack of value is not an error, but shouldn't it be one here? If consensus asks for validators and contract runtime can't provide them, something somewhere went wrong, so I think such a case could (and maybe even should) be considered an error.

Reply:
> If consensus asks for validators and contract runtime can't provide them, something somewhere went wrong
The `Ok(None)` case means simply everything went ok, but the data for given era is not yet computed. I'd agree though if the passed era should be validated - then returning an error makes perfect sense.

Reply:
I think that's the case here - we should never be getting the validators for an era that hasn't been computed yet. Not 100% sure though, feel free to wait if someone else confirms.

Reply:
As discussed today, it's okay the way it is :+1: